plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+	session->media.video_srtp_in = NULL;
+	g_free(session->media.video_remote_policy.key);
+	session->media.video_remote_policy.key = NULL;
+	session->media.video_srtp_suite_in = 0;
+}
+
+
+/* SDP parsing and manipulation */
+void janus_nosip_sdp_process(janus_nosip_session *session, janus_sdp *sdp, gboolean answer, gboolean update, gboolean *changed);
+char *janus_nosip_sdp_manipulate(janus_nosip_session *session, janus_sdp *sdp, gboolean answer);
+/* Media */
+static int janus_nosip_allocate_local_ports(janus_nosip_session *session);
+static void *janus_nosip_relay_thread(void *data);
+
+
+/* Random string helper (in case user doesn't provide opaque info) */
+static char charset[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+static void janus_nosip_random_string(int length, char *buffer) {
+	if(length > 0 && buffer) {
+		int l = (int)(sizeof(charset)-1);
+		int i=0;
+		for(i=0; i<length; i++) {
+			int key = rand() % l;
+			buffer[i] = charset[key];
+		}
+		buffer[length-1] = '\0';
+	}
+}
+
+
+/* Error codes */
+#define JANUS_NOSIP_ERROR_UNKNOWN_ERROR			499
+#define JANUS_NOSIP_ERROR_NO_MESSAGE			440
+#define JANUS_NOSIP_ERROR_INVALID_JSON			441
+#define JANUS_NOSIP_ERROR_INVALID_REQUEST		442
+#define JANUS_NOSIP_ERROR_MISSING_ELEMENT		443
+#define JANUS_NOSIP_ERROR_INVALID_ELEMENT		444
+#define JANUS_NOSIP_ERROR_WRONG_STATE			445
+#define JANUS_NOSIP_ERROR_MISSING_SDP			446
+#define JANUS_NOSIP_ERROR_INVALID_SDP			447
+#define JANUS_NOSIP_ERROR_IO_ERROR				448
+#define JANUS_NOSIP_ERROR_RECORDING_ERROR		449
+#define JANUS_NOSIP_ERROR_TOO_STRICT			450
+
+
+/* NoSIP watchdog/garbage collector (sort of) */
+static void *janus_nosip_watchdog(void *data) {
+	JANUS_LOG(LOG_INFO, "NoSIP watchdog started\n");
+	gint64 now = 0;
+	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+		janus_mutex_lock(&sessions_mutex);
+		/* Iterate on all the sessions */
+		now = janus_get_monotonic_time();
+		if(old_sessions != NULL) {
+			GList *sl = old_sessions;
+			JANUS_LOG(LOG_HUGE, "Checking %d old NoSIP sessions...\n", g_list_length(old_sessions));
+			while(sl) {
+				janus_nosip_session *session = (janus_nosip_session *)sl->data;
+				if(!session) {
+					sl = sl->next;
+					continue;
+				}
+				if (now-session->destroyed >= 5*G_USEC_PER_SEC) {
+					/* We're lazy and actually get rid of the stuff only after a few seconds */
+					JANUS_LOG(LOG_VERB, "Freeing old NoSIP session\n");
+					GList *rm = sl->next;
+					old_sessions = g_list_delete_link(old_sessions, sl);
+					sl = rm;
+					g_free(session->info);
+					session->info = NULL;
+					janus_sdp_free(session->sdp);
+					session->sdp = NULL;
+					g_free(session->media.remote_ip);
+					session->media.remote_ip = NULL;
+					janus_nosip_srtp_cleanup(session);
+					session->handle = NULL;
+					g_free(session);
+					session = NULL;
+					continue;
+				}
+				sl = sl->next;
+			}
+		}
+		janus_mutex_unlock(&sessions_mutex);
+		g_usleep(500000);
```

Comment:
That's the way all plugins behave right now, but it has already been dealt with in the referece counters branch, #403, where we don't have that anymore.